File: backend/app/calculation/simulator.py
"""興行シミュレーション。DemandEstimatorの出力（需要）とキャパ・会場費から
売上・利益・稼働率を計算する。DemandEstimatorの具象実装には依存しない。
"""
from __future__ import annotations

from app.calculation.demand_estimator import DemandEstimator
from app.calculation.types import DemandFeatures, ScenarioResult, VenueCandidate
from app.calculation.venue_fit import classify_venue_fit


class PerformanceSimulator:
    def __init__(self, demand_estimator: DemandEstimator):
        self._demand_estimator = demand_estimator

    def simulate(
        self,
        features: DemandFeatures,
        venue: VenueCandidate,
        price: int,
        num_performances: int,
    ) -> ScenarioResult:
        scenario_features = DemandFeatures(
            group=features.group,
            past_performances=features.past_performances,
            current_production=features.current_production,
            venue=venue,
            price=price,
            num_performances=num_performances,
        )
        estimate = self._demand_estimator.estimate_demand(scenario_features)

        available_seats = venue.capacity * num_performances
        expected_sold = min(available_seats, estimate.total_expected_demand)
        occupancy_rate = expected_sold / available_seats if available_seats > 0 else 0.0
        revenue = expected_sold * price
        profit = revenue - venue.venue_cost * num_performances

        venue_fit, venue_fit_message = classify_venue_fit(occupancy_rate)

        return ScenarioResult(
            venue_name=venue.name,
            price=price,
            num_performances=num_performances,
            available_seats=available_seats,
            expected_demand=estimate.total_expected_demand,
            expected_sold=expected_sold,
            occupancy_rate=occupancy_rate,
            revenue=revenue,
            profit=profit,
            venue_fit=venue_fit,
            venue_fit_message=venue_fit_message,
        )

File: frontend/src/calculation/simulator.ts
/** rule_v0.1: backend/app/calculation/simulator.py の移植。 */
import { estimateDemand } from "./demandEstimator";
import type { DemandFeatures, ScenarioResult } from "./types";
import type { VenueCandidate } from "../types";
import { classifyVenueFit } from "./venueFit";

export function simulate(
  features: DemandFeatures,
  venue: VenueCandidate,
  price: number,
  numPerformances: number
): ScenarioResult {
  const scenarioFeatures: DemandFeatures = {
    group: features.group,
    past_performances: features.past_performances,
    current_production: features.current_production,
    venue,
    price,
    num_performances: numPerformances,
  };
  const estimate = estimateDemand(scenarioFeatures);

  const availableSeats = venue.capacity * numPerformances;
  const expectedSold = Math.min(availableSeats, estimate.total_expected_demand);
  const occupancyRate = availableSeats > 0 ? expectedSold / availableSeats : 0.0;
  const revenue = expectedSold * price;
  const profit = revenue - venue.venue_cost * numPerformances;

  const [venueFit, venueFitMessage] = classifyVenueFit(occupancyRate);

  return {
    venue_name: venue.name,
    price,
    num_performances: numPerformances,
    available_seats: availableSeats,
    expected_demand: estimate.total_expected_demand,
    expected_sold: expectedSold,
    occupancy_rate: occupancyRate,
    revenue,
    profit,
    venue_fit: venueFit,
    venue_fit_message: venueFitMessage,
  };
}
